include/deal.II/non_matching/mapping_info.h

```diff
@@ -157,30 +293,550 @@ namespace NonMatching
+                                                 numbers::invalid_unsigned_int);
+    cell_index = 0;
+    for (const auto &cell : cell_iterator_range)
+      {
+        auto it = unit_points.begin() + unit_points_index[cell_index];
+        for (const auto &unit_point : unit_points_vector[cell_index])
+          {
+            *it = unit_point;
+            ++it;
+          }
+
+        compute_mapping_data_for_generic_points(cell,
+                                                unit_points_vector[cell_index],
+                                                mapping_data[cell_index]);
+
+        if (do_cell_index_compression)
+          cell_index_to_compressed_cell_index[cell->active_cell_index()] =
+            cell_index;
+
+        ++cell_index;
+      }
+
+    state = State::cell_vector;
+  }
+
+
+
+  template <int dim, int spacedim>
+  template <typename Iterator>
+  void
+  MappingInfo<dim, spacedim>::reinit_cells(
+    const IteratorRange<Iterator> &     cell_iterator_range,
+    const std::vector<Quadrature<dim>> &quadrature_vector,
+    const unsigned int                  n_unfiltered_cells)
+  {
+    const unsigned int n_cells =
+      std::distance(cell_iterator_range.begin(), cell_iterator_range.end());
+
+    Assert(n_cells == quadrature_vector.size(),
+           ExcDimensionMismatch(n_cells, quadrature_vector.size()));
+
+    std::vector<std::vector<Point<dim>>> unit_points_vector(n_cells);
+    for (unsigned int cell_index = 0; cell_index < n_cells; ++cell_index)
+      unit_points_vector[cell_index] = std::vector<Point<dim>>(
+        quadrature_vector[cell_index].get_points().begin(),
+        quadrature_vector[cell_index].get_points().end());
+
+    reinit_cells(cell_iterator_range, unit_points_vector, n_unfiltered_cells);
+
+    if (update_flags_mapping & update_JxW_values)
+      for (unsigned int cell_index = 0; cell_index < n_cells; ++cell_index)
+        {
+          const auto &weights = quadrature_vector[cell_index].get_weights();
+          for (unsigned int q = 0; q < weights.size(); ++q)
+            mapping_data[cell_index].JxW_values[q] =
+              determinant(
+                Tensor<2, dim>(mapping_data[cell_index].jacobians[q])) *
+              weights[q];
+        }
+  }
+
+
+
+  template <int dim, int spacedim>
+  template <typename Iterator>
+  void
+  MappingInfo<dim, spacedim>::reinit_surface(
+    const IteratorRange<Iterator> &                    cell_iterator_range,
+    const std::vector<ImmersedSurfaceQuadrature<dim>> &quadrature_vector,
+    const unsigned int                                 n_unfiltered_cells)
+  {
+    do_cell_index_compression =
+      n_unfiltered_cells != numbers::invalid_unsigned_int;
+
+    if (update_flags_mapping & (update_JxW_values | update_normal_vectors))
+      update_flags_mapping |= update_covariant_transformation;
+
+    const unsigned int n_cells =
+      std::distance(cell_iterator_range.begin(), cell_iterator_range.end());
```

Comment:
Since this and the following two functions are pretty similar, I wonder whether you see a way to implement the algorithm only once and use some unifying calls (e.g. transformation to arguments needed by other function) instead?

Reply:
The two `reinit_faces` are pretty similar indeed. I would suggest removing the `reinit_faces` function that only takes vectors of unit points on faces. I do not use this function anymore since I implemented the function that takes quadratures and it is also not tested by the test in this PR. What do you think?

Reply:
> I would suggest removing the reinit_faces function that only takes vectors of unit points on faces. I do not use this function anymore since I implemented the function that takes quadratures and it is also not tested by the test in this PR. What do you think?

I agree, that would help.

Reply:
Done.